torch/utils/data/dataset.py

```diff
@@ -189,6 +191,21 @@ def set_reduce_ex_hook(cls, hook_fn):
             raise Exception("Attempt to override existing reduce_ex_hook")
         IterableDataset.reduce_ex_hook = hook_fn
 
+    #  def __init_subclass__(cls, *args, **kwargs):
+    #      # TODO: Determinin if force all IterableDataset and IterDataPipe with annotation
+    #      if not hasattr(cls, 'type'):
+    #          raise TypeError('Class {} needs to be specified with type'.format(cls.__name__))
+
+    @_tp_cache
+    def __class_getitem__(cls, param) -> _DataPipeAlias:
+        # TODO: add global switch for type checking at compile-time
+        if param is None:
+            raise TypeError('Can not take None as type parameter for {}'.format(cls.__name__))
+        if isinstance(param, Sequence):
+            param = Tuple[param]
+        _type_check(param, msg="Parameters for {} must be types".format(cls.__name__))
+        return _DataPipeAlias(cls, param)
+
```

Comment:
This part can be moved to `DataSet` to make all previous map-stype Dataset gains the benefits (?) from more strict typing requirement.